jms/src/main/scala/akka/stream/alpakka/jms/Jms.scala

```diff
@@ -58,13 +61,52 @@ object AcknowledgeMode {
   val SessionTransacted: AcknowledgeMode = new AcknowledgeMode(jms.Session.SESSION_TRANSACTED)
 }
 
+object ConnectionRetrySettings {
+  def create(): ConnectionRetrySettings = ConnectionRetrySettings()
+}
+
+final case class ConnectionRetrySettings(connectTimeout: FiniteDuration = 10.seconds,
+                                         initialRetry: FiniteDuration = 100.millis,
+                                         backoffFactor: Double = 2,
+                                         maxBackoff: FiniteDuration = 1.minute,
+                                         maxRetries: Int = 10) {
+  def withConnectTimeout(timeout: FiniteDuration): ConnectionRetrySettings = copy(connectTimeout = timeout)
+  def withConnectTimeout(timeout: Long, unit: TimeUnit): ConnectionRetrySettings =
+    copy(connectTimeout = Duration(timeout, unit))
+  def withInitialRetry(delay: FiniteDuration): ConnectionRetrySettings = copy(initialRetry = delay)
+  def withInitialRetry(delay: Long, unit: TimeUnit): ConnectionRetrySettings =
+    copy(initialRetry = Duration(delay, unit))
+  def withBackoffFactor(backoffFactor: Double): ConnectionRetrySettings = copy(backoffFactor = backoffFactor)
+  def withMaxBackoff(maxBackoff: FiniteDuration): ConnectionRetrySettings = copy(maxBackoff = maxBackoff)
+  def withMaxBackoff(maxBackoff: Long, unit: TimeUnit): ConnectionRetrySettings =
+    copy(maxBackoff = Duration(maxBackoff, unit))
+  def withMaxRetries(maxRetries: Int): ConnectionRetrySettings = copy(maxRetries = maxRetries)
+
+  /** Hypothetical retry time, not accounting for maxBackoff. */
+  def waitTime(retryNumber: Int): FiniteDuration =
+    (initialRetry * Math.pow(retryNumber, backoffFactor)).asInstanceOf[FiniteDuration]
+
+  /** Max possible time it will take to timeout, including connection timeouts and the wait time between retries.  */
+  def maxWaitTime: Duration =
+    if (maxRetries < 0) Duration.Inf
+    else {
+      val totalWaitTime = (0 to maxRetries).map(connectTimeout + waitTime(_)).reduce(_ + _)
+      val totalTimeout = connectTimeout * (maxRetries + 1)
+      totalTimeout + totalWaitTime
+    }
+
+}
+
+case class ConnectionRetryException(message: String, cause: Throwable) extends Exception(message, cause)
```

Comment:
how about moving that exception to `JmsExceptions.scala`? I think either all exceptions should live there, or all live in `Jms.scala`

Reply:
Sure. Will do.